### PR TITLE
(PUP-7258) Use native JSON parser and generator

### PR DESCRIPTION
**Pull request by joshcooper**
Commit 30e0f1bcd was supposed to remove the json_pure dependency, but
due to a conflict when rebasing wasn't. As a result, puppet was using
json_pure instead of native json in some cases. It also exposed other
places where json_pure was being used:

* Bundler within puppet uses Gemfile, which includes .gemspec
* Bundler for a module that depends on a source checkout of puppet uses .gemspec
* Packaging uses project_data.yaml when building puppet as a gem
* Git acceptance on Solaris (10 and 11) install json_pure to satisfy
  puppet's dependencies.

All traces of json_pure have been removed so that the native version
already present in ruby 1.9.3 and up is used.

### CODE REVIEW

**Comment by joshcooper** [comment]
In puppet:

    $ bundle exec ruby -e "require 'json'; puts JSON::Parser"
    JSON::Ext::Parser

In puppetlabs-apt:

    $ env PUPPET_GEM_VERSION=file:///Users/josh/work/puppet bundle install --path .bundle --without system_tests
    ...
    $ env PUPPET_GEM_VERSION=file:///Users/josh/work/puppet bundle exec ruby -e "require 'json'; puts JSON::Parser"
    JSON::Ext::Parser

Installing puppet as a gem:
    root@ejv3yz9944qea6p:~# gem install puppet-4.10.0.433.dirty.gem
    Fetching: locale-2.1.2.gem (100%)
    Successfully installed locale-2.1.2
    Fetching: fast_gettext-1.1.0.gem (100%)
    Successfully installed fast_gettext-1.1.0
    Fetching: text-1.3.1.gem (100%)
    Successfully installed text-1.3.1
    Fetching: gettext-3.2.2.gem (100%)
    Successfully installed gettext-3.2.2
    Fetching: gettext-setup-0.21.gem (100%)
    Successfully installed gettext-setup-0.21
    Fetching: hiera-3.3.1.gem (100%)
    Successfully installed hiera-3.3.1
    Fetching: facter-2.4.6.gem (100%)
    Successfully installed facter-2.4.6
    Successfully installed puppet-4.10.0.433.dirty
    ...
    root@ejv3yz9944qea6p:~# ruby -e "require 'json'; puts JSON::Parser"
    JSON::Ext::Parser

On Solaris 10:

    root@yk8rsu4mccvdefb:~# gem which json
    /opt/csw/lib/ruby/1.9.1/json.rb
    root@yk8rsu4mccvdefb:~# ruby -e "require 'json'; puts JSON::Parser"
    JSON::Ext::Parser

**Comment by Iristyle** [comment]
Is there a corresponding puppet-agent PR?

File: benchmarks/serialization/benchmarker.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  def generate
+    path = File.expand_path(File.join(__FILE__, '../catalog.json'))
+    puts "Using catalog #{path}"
+
+    @data = File.read(path)
+    @catalog = JSON.parse(@data)
+  end
+
+  def run(args=nil)
+    0.upto(ENV['ITERATIONS'].to_i) do |i|
```
Comment (MosesMendoza) [inline benchmarks/serialization/benchmarker.rb]: I think the benchmark run task [already iterates over `ENV['ITERATIONS']`](https://github.com/puppetlabs/puppet/blob/master/tasks/benchmark.rake#L43) - just confirming we want to separately iterate up to the count here, ie ultimately producing parse/generate results for ENV['ITERATIONS']^2 iterations?

Comment (joshcooper) [inline benchmarks/serialization/benchmarker.rb]: Good point, I'll fix that. I realized I have the wrong ticket number on some of the commits and branch name. I'm going to close this PR and reopen a new one.

**Pull request closed without merging**